app/renderer/actions/Session.js

```diff
@@ -572,6 +582,9 @@ export function newSession(caps, attachSessId = null) {
         let attachedSessionCaps = {};
         if (attachedSession) {
           attachedSessionCaps = attachedSession.capabilities;
+          if (attachedSessionCaps.platformName === undefined) {
```

Comment:
could you please describe the purpose of this change?

Reply:
When attaching to w3c sessions, capabilities are nested inside the values. As per the current logic, all the values are mapped as capabilities and in further lines, we set `serverOpts.isIOS` and `serverOpts.isAndroid` by performing regex match on this. Since in w3c session, capabilities are nested in values, appium inspector was giving `cannot read property match of undefined` error and attach to session fails.

This piece of code checks if platformName exists in attachedSessionCaps if not assign the capabilities nested inside.

https://github.com/appium/appium-inspector/blob/75baa6ed3a6e8eedfa2d6c1a099a09e55f36537b/app/renderer/actions/Session.js#L594

Reply:
session data for reference
`{
    "value": {
        "capabilities": {
            "adbExecTimeout": 120000,
            "allowInvisibleElements": true,
            "autoAcceptAlerts": true,
            "autoGrantPermissions": true,
            "automationName": "uiautomator2",
            "chromeOptions": {
                "args": []
            },
            "desired": {
                "adbExecTimeout": 120000,
                "allowInvisibleElements": true,
                "autoAcceptAlerts": true,
                "autoGrantPermissions": true,
                "automationName": "uiautomator2",
                "chromeOptions": {
                    "args": []
                },
                "headless": false,
                "nativeWebScreenshot": true,
                "newCommandTimeout": 0,
                "orientation": "PORTRAIT",
                "platformName": "android",
                "udid": "emulator-5554",
                "uiautomator2ServerLaunchTimeout": 60000,
                "waitForQuiescence": false
            },
            "deviceScreenDensity": 320,
            "deviceScreenSize": "1080x2400",
            "headless": false,
            "javascriptEnabled": true,
            "locationContextEnabled": false,
            "nativeWebScreenshot": true,
            "networkConnectionEnabled": true,
            "newCommandTimeout": 0,
            "orientation": "PORTRAIT",
            "pixelRatio": 2,
            "platformName": "android",
            "platformVersion": "13",
            "statBarHeight": 48,
            "takesScreenshot": true,
            "uiautomator2ServerLaunchTimeout": 60000,
            "viewportRect": {
                "height": 2304,
                "left": 0,
                "top": 48,
                "width": 1080
            },
            "waitForQuiescence": false,
            "warnings": {},
            "webStorageEnabled": false
        }
        "sessionId": ""
    }
}`

Reply:
@eglitise does it ring a bell for you? Also I suppose this is supposed to fix an issue which is different from the one this PR tries to address and should probably be addressed separately.

Reply:
I'll take a look tonight further but I thought below code returned info in `value`, then the `attachedSession.capabilities` may have `attachedSession.capabilities.platformName`. If not, probably the response was not expected one(?)
```
const attachedSession = session.runningAppiumSessions.find(
          (session) => session.id === attachSessId,
        );
```

As eglitise's comment, if the response included value, then `attachedSession.value.capabilities` would be actually reasonable as a fix.

Reply:
Hm, checked with the local Appium for the `{"value": {"capabilities": {....}}}`.

`attachedSession`:
![Screenshot 2024-03-02 at 1 22 06 AM](https://github.com/appium/appium-inspector/assets/5511591/043d2e9b-2d77-4eac-a132-f1a44bb39fc7)


`attachedSessionCaps`:
![Screenshot 2024-03-02 at 1 24 22 AM](https://github.com/appium/appium-inspector/assets/5511591/56dbdc32-42e3-4762-9d55-65844de113eb)

@shiva-guntoju Could you share the entire value of `attachedSession` you observed? 

I wondered if the `attachedSession` has additional `capabilities` like `{"value": {"session_id": "....", "capabilities": {"capabilities": ...}}}`. If it did not have such additional `capabilities`, I guess `attachedSession.capabilities.capabilities` would be `undefined`.

Reply:
@eglitise @KazuCocoa @mykola-mokhnach these are the two formats I have observed.
```javascript
//JSON wire session data:
{
    "sessionId": "09d62aee-d233-4bcb-82a9-72ae5e2e231f",
    "status": 0,
    "value": {
        "adbExecTimeout": 120000,
        "autoAcceptAlerts": true,
        "automationName": "uiautomator2",
        "chromeOptions": {
            "args": [],
            "w3c": false
        },
        "chromedriverPort": 10314,
        "databaseEnabled": false,
        "desired": {
            "adbExecTimeout": 120000,
            "autoAcceptAlerts": true,
            "automationName": "uiautomator2",
            "chromeOptions": {
                "args": [],
                "w3c": false
            },
            "headless": false,
            "nativeWebScreenshot": true,
            "newCommandTimeout": 0,
            "orientation": "PORTRAIT",
            "platformName": "android",
            "platformVersion": "10",
            "skipLogcatCapture": true,
            "uiautomator2ServerLaunchTimeout": 60000,
            "waitForQuiescence": false
        },
        "platformVersion": "10",
        "skipLogcatCapture": true,
        "statBarHeight": 80,
        "systemPort": 12145,
        "takesScreenshot": true,
        "udid": "8e60453b",
        "uiautomator2ServerLaunchTimeout": 60000,
        "viewportRect": {
            "height": 2138,
            "left": 0,
            "top": 80,
            "width": 1080
        },
        "waitForQuiescence": false,
        "warnings": {},
        "webStorageEnabled": false
    }
}
``` 

```javascript
//w3c session data:
{
  "value": {
    "capabilities": {
      "adbExecTimeout": 120000,
      "allowInvisibleElements": true,
      "autoAcceptAlerts": true,
      "autoGrantPermissions": true,
      "automationName": "uiautomator2",
      "chromeOptions": {
        "args": []
      },
      "desired": {
        "adbExecTimeout": 120000,
        "allowInvisibleElements": true,
        "autoAcceptAlerts": true,
        "autoGrantPermissions": true,
        "automationName": "uiautomator2",
        "chromeOptions": {
          "args": []
        },
        "headless": false,
        "nativeWebScreenshot": true,
        "newCommandTimeout": 0,
        "orientation": "PORTRAIT",
        "platformName": "android",
        "udid": "emulator-5554",
        "uiautomator2ServerLaunchTimeout": 60000,
        "waitForQuiescence": false
      },
      "deviceScreenDensity": 320,
      "deviceScreenSize": "1080x2400",
      "headless": false,
      "javascriptEnabled": true,
      "locationContextEnabled": false,
      "nativeWebScreenshot": true,
      "networkConnectionEnabled": true,
      "newCommandTimeout": 0,
      "orientation": "PORTRAIT",
      "pixelRatio": 2,
      "platformName": "android",
      "platformVersion": "13",
      "statBarHeight": 48,
      "takesScreenshot": true,
      "uiautomator2ServerLaunchTimeout": 60000,
      "viewportRect": {
        "height": 2304,
        "left": 0,
        "top": 48,
        "width": 1080
      },
      "waitForQuiescence": false,
      "warnings": {},
      "webStorageEnabled": false
    },
    "sessionId": ""
  }
}
``` 

Current flow works as expected for JSON wire data as capabilities passed for session creation are stored for `value` key which is stored as `attachedSession.capabilities` I could not find the exact place where we are mapping value to capabilities but it is working for json wire sessions. My explanation is based on this and I will look further for the piece of code which does this.

And incase of w3c sessions, since we are already storing data of `value` in original session data as `attachedSession.capabilities`, the nested capabilities which is inside the `value` in original response is now under `attachedSession.capabilties.capabilities`. 

We have tested the new changes by attaching to both josnwire and w3c sessions and we have successfully attached to the session. 

Hope this answers all the questions and let me know incase of further questions.

Reply:
Thank you.

I was a bit wrong, sorry. what we should check here is `/sessions` endpoint, which is not W3C spec but defined in Appium as https://appium.io/docs/en/latest/commands/base-driver/#getsessions, correct? (and the format actually hasn't been changed since it is not in W3C). It is not the response of new session request nor get session endpoint.

So, the `sessions`'s endpoint's response would be formatted as:

```
{ 
  "value": [
    {"session_id": "... session 1", "capabilities": {....}},
    {"session_id": "... session 2", "capabilities": {....}}
  ]
}
```

`runningAppiumSessions` used by `attachedSession.capabilities` in:
```
        const attachedSession = session.runningAppiumSessions.find(
          (session) => session.id === attachSessId,
        );
```

https://github.com/appium/appium-inspector/blob/f4507409928284b0cd6c1bdaa231d50e31e456f7/app/renderer/reducers/Session.js#L296


comes from:

https://github.com/appium/appium-inspector/blob/f4507409928284b0cd6c1bdaa231d50e31e456f7/app/renderer/actions/Session.js#L945

```
      console.log(`sessions: ----> ${JSON.stringify(res.data.value)}`);
      dispatch({type: GET_SESSIONS_DONE, sessions: res.data.value});
```

The `res.data.value` might be like:
```
sessions: ----> [{"id":"61a63c8b-96eb-419d-8002-c41cd235973b","capabilities":{"platformName":"ios","automationName":"xcuitest","platformVersion":"17.2","deviceName":"iPhone 15","webDriverAgentUrl":"http://localhost:8100"}}]
```

Could you share your `/sessions` endpoint call as well?

Now, I am wondering if your response value is like below and it has extra `capabilities`.

```
{ 
  "value": [
    {"session_id": "... session 1", "capabilities": {"capabilities": {....}}
  ]
}
```

Reply:
What I could share right now is please make sure the response follows https://appium.io/docs/en/latest/commands/base-driver/#getsessions format as the endpoint. Possibly your service should have an orchestration layer to modify the response.

Reply:
> Btw, the Appium/WDA is expected to be called by the Appium XCUITest driver in our current design. The direct response by WDA might not fit our Appium XCUITest driver response.

I agree on this and we follow the same, since WDA is the one that handles all the commands, I mentioned it as the response from WDA. 

> Possibly your service might need to have an orchestration layer to modify the response.

This is definitely one way of doing it but the reason we chose not to do it is that if we translate the session response from w3c to jsonwire we would have to translate the subsequent commands from appium inspector from jsonwire to w3c, which adds a significant overhead. We send the raw session response to appium inspector and let appium inspector decide the commands based on the session. We are open for suggestions here.

Reply:
Oh, so, you're suggesting a new specification for the `/sessions` endpoint, instead of current drivers' response, correct? (to detect which session is w3c, or mjsonwp for your platform)

For example, the latest xcuitest driver's response of `/sessions` endpoint follows https://appium.io/docs/en/latest/commands/base-driver/#getsessions so the response of `/sessions` is like:
```
{
    "value": [
        {
            "id": "fafef641-1d67-4460-95e8-0dd8c2bff7e6",
            "capabilities": {
                "platformName": "ios",...
            }
        }
    ]
}
``` 

but you're suggesting a new format to have nested `capabilities` like:

```
{
    "value": [
        {
            "id": "fafef641-1d67-4460-95e8-0dd8c2bff7e6",
            "capabilities": {
                "capabilities": {
                    "platformName": "ios",...
                }
            }
        }
    ]
}
```

which has nested `capabilities` key in case the session is w3c

Reply:
I agree with the suggestions and we were unaware of the discussions, thanks for sharing them.

As suggested, we will make the change to include this only for our platform.

Reply:
Will raise a new PR with the suggested changes, closing this for now.